README.md
# pyTarget
iSCSI target for Python 2.7

This is a fork of original pyTarget project from SourceForge
https://sourceforge.net/projects/pytarget/


## Additional features compared to original:
- Ability to use physical disks as data source
- Ability to auto-detect and grow disk capacity
- Ability to use volumes (with emulated GPT) as data source
- 2TB+ disks support
- Customize disk physical and logical sector size

## Prerequisites
- Python 2.7.10

### Windows only
- Install `pywin32` module by running cmd.exe as admin and typing `C:\Python27\python -m pip install pywin32`
- To access physical disks and volumes, pyTarget must be run as admin

#### Connect to a local physical disk
1. Open cmd.exe as admin, start iscsi initiator by running 
- `sc config msiscsi start= auto`
- `net start msiscsi`
2. Get iscsi intiator address e.g. `iqn.1991-05.com.microsoft:vlad-test-2019` by:
- `echo iqn | iscsicli | findstr [iqn`
3. Locate physical disk size in bytes and path in form of `\\.\PhysicalDriveX` with `wmic diskdrive get Manufacturer, DeviceId, Model, Index, Size`
4. Open `pytarget/src/config.xml` and insert the following line inside of `<target>` tag:
- `<host name="IQN" target_pwd="" initiator_pwd="> <lun id="0" type="0" path="\\.\PhysicalDrive2" capacity="6442450944" media="0x1000" lsector="512" psector="4096"/> </host>`
- replace `IQN` in `name` attribute with with one you got from step earlier
- replace `\\.\PhysicalDrive2` in `path` with windows drive path
- for fixed capacity, replace `capacity` value with size in sectors, for that divide size in bytes by 512; for dynamic capacity set the value to "0" for auto-detect and auto-expand
- fix 'psector' value if you want to override physical sector size reported by iscsi target

Example, in case we want to use disk #3:
```
C:\Users\feoff\Desktop\pytarget>echo iqn | iscsicli | findstr [iqn
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit

C:\Users\feoff\Desktop\pytarget>wmic diskdrive get Manufacturer, DeviceId, Model, Index, Size
DeviceID            Index  Manufacturer            Model                   Size
\\.\PHYSICALDRIVE0  0      (Standard disk drives)  Microsoft Virtual Disk  136366917120
\\.\PHYSICALDRIVE1  1      (Standard disk drives)  Microsoft Virtual Disk  7509680640
\\.\PHYSICALDRIVE2  2      (Standard disk drives)  Microsoft Virtual Disk  4293596160
\\.\PHYSICALDRIVE3  3      (Standard disk drives)  Microsoft Virtual Disk  34356994560
```

Then the corresponding entry in config.xml will look like
```
...
    <target name="iqn.2006-11.1.python.iscsi.target-1" ip="127.0.0.1" port="3260" portal="1">
    
    <host name="iqn.1991-05.com.microsoft:vlad-test-2019" target_pwd="" initiator_pwd=""> <lun id="0" type="0" path="\\.\PHYSICALDRIVE3" capacity="67103505" media="0x1000"/> </host>
 
...
```

5. Start the target:
- Open `cmd.exe` and navigate to pytarget folder, `src` subfolder
- `C:\Python27\python.exe pyTarget.py`
6. Start iscsi intiator to connect to target, from new cmd.exe console (change `127.0.0.1` to target IP address if it is installed on other PC in the network):
- `iscsicli QAddTargetPortal 127.0.0.1`
- `iscsicli qlogintarget iqn.2006-11.1.python.iscsi.target-1`
- Check the disk is connected by `iscsicli reporttargetmappings`

7. To end the session (and disconnect the disk):
- `iscsicli logouttarget <session-id>`, where <session-id>  is reported by `iscsicli reporttargetmappings`, e.g. `iscsicli logouttarget fffffa800626e018-4000013700000006`

#### Connect to a local volume
1. Open cmd.exe as admin, start iscsi initiator by running 
- `sc config msiscsi start= auto`
- `net start msiscsi`
2. Get iscsi intiator address e.g. `iqn.1991-05.com.microsoft:vlad-test-2019` by:
- `echo iqn | iscsicli | findstr [iqn`
3. Locate basic or dynamic volume guid path `\\?\Volume{GUID}` as returned by `mountvol X: /L`, replace X: with the target volume letter
4. Open `pytarget/src/config.xml` and insert the following line inside of `<target>` tag:
- `<host name="IQN" target_pwd="" initiator_pwd="> <lun id="0" type="0" path="\\?\Volume{GUID}" capacity="6442450944" media="0x2000" lsector="512" psector="4096"/> </host>`
- replace `IQN` in `name` attribute with with one you got from step earlier
- replace `\\?\Volume{GUID}` in `path` with volume guid path (no trailing slash)
- for fixed capacity, replace `capacity` value with size in sectors, for that divide size in bytes by 512; for dynamic capacity set the value to "0" for auto-detect and auto-expand
- fix 'psector' value if you want to override physical sector size reported by iscsi target

Example, in case we want to use volume F:
```
C:\Users\feoff\Desktop\pytarget>echo iqn | iscsicli | findstr [iqn
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit

C:\Users\feoff\Desktop\pytarget>mountvol F:\ /L
    \\?\Volume{9bb6d08d-6370-11ee-a534-6045bdad03c5}\
```

Then the corresponding entry in config.xml will look like
```
...
    <target name="iqn.2006-11.1.python.iscsi.target-1" ip="127.0.0.1" port="3260" portal="1">
    
    <host name="iqn.1991-05.com.microsoft:vlad-test-2019" target_pwd="" initiator_pwd=""> <lun id="0" type="0" path="\\?\Volume{9bb6d08d-6370-11ee-a534-6045bdad03c5}" capacity="0" media="0x2000"/> </host>
 
...
```

5. Start the target:
- Open `cmd.exe` and navigate to pytarget folder, `src` subfolder
- `C:\Python27\python.exe pyTarget.py`
6. Start iscsi intiator to connect to target, from new cmd.exe console (change `127.0.0.1` to target IP address if it is installed on other PC in the network):
- `iscsicli QAddTargetPortal 127.0.0.1`
- `iscsicli qlogintarget iqn.2006-11.1.python.iscsi.target-1`
- Check the disk is connected by `iscsicli reporttargetmappings`
7. To end the session (and disconnect the disk):
- `iscsicli logouttarget <session-id>`, where <session-id>  is reported by `iscsicli reporttargetmappings`, e.g. `iscsicli logouttarget fffffa800626e018-4000013700000006`


#### Connect to a local volume, emulate GPT layout
1. Open cmd.exe as admin, start iscsi initiator by running 
- `sc config msiscsi start= auto`
- `net start msiscsi`
2. Get iscsi intiator address e.g. `iqn.1991-05.com.microsoft:vlad-test-2019` by:
- `echo iqn | iscsicli | findstr [iqn`
3. Generate footer and header file, 32MB size: `fsutil file createnew C:\footer 33554432`, `fsutil file createnew C:\header 33554432`
4. Locate basic or dynamic volume guid path `\\?\Volume{GUID}` as returned by `mountvol X: /L`, replace X: with the target volume letter
5. Open `pytarget/src/config.xml` and insert the following line inside of `<target>` tag:
- `<host name="IQN" target_pwd="" initiator_pwd="> <lun id="0" type="0" path="\\?\Volume{GUID}" capacity="6442450944" media="0x4000" lsector="512" psector="4096" parms="C:\header;C:\footer"/> </host>`
- replace `IQN` in `name` attribute with with one you got from step earlier
- replace `\\?\Volume{GUID}` in `path` with volume guid path (no trailing slash)
- for fixed capacity, replace `capacity` value with size in sectors, for that divide size in bytes by 512; for dynamic capacity set the value to "0" for auto-detect and auto-expand
- fix 'psector' value if you want to override physical sector size reported by iscsi target
- in 'parms' field two preallocated files must be specified: one that will have starting sectors of emulated GPT, and one that will have the ending sectors of emulated GPT

Example, in case we want to use volume F:
```
C:\Users\feoff\Desktop\pytarget>echo iqn | iscsicli | findstr [iqn
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit
[iqn.1991-05.com.microsoft:vlad-test-2019] Enter command or ^C to exit

C:\Users\feoff\Desktop>fsutil file createnew C:\header 33554432
File C:\header is created

C:\Users\feoff\Desktop>fsutil file createnew C:\footer 33554432
File C:\footer is created

C:\Users\feoff\Desktop\pytarget>mountvol F:\ /L
    \\?\Volume{9bb6d08d-6370-11ee-a534-6045bdad03c5}\

```

Then the corresponding entry in config.xml will look like
```
...
    <target name="iqn.2006-11.1.python.iscsi.target-1" ip="127.0.0.1" port="3260" portal="1">
    
    <host name="iqn.1991-05.com.microsoft:vlad-test-2019" target_pwd="" initiator_pwd=""> <lun id="0" type="0" path="\\?\Volume{9bb6d08d-6370-11ee-a534-6045bdad03c5}" capacity="0" media="0x4000" parms="C:\header;C:\footer" psector="4096"/> </host>
 
...
```

5. (Optionally) If some of the apps are using the source volume
- TODO: describe how to change the mount points 
- Restart
- TODO: describe how to add the target to autostart (maybe to service with early loading group)

6. Start the target:
- Open `cmd.exe` and navigate to pytarget folder, `src` subfolder
- `C:\Python27\python.exe pyTarget.py`
- Read 'main dev size: ' value in console (it will show the size of partition to create in bytes)

8. Start iscsi intiator to connect to target, from new cmd.exe console (change `127.0.0.1` to target IP address if it is installed on other PC in the network):
- `iscsicli QAddTargetPortal 127.0.0.1`
- `iscsicli qlogintarget iqn.2006-11.1.python.iscsi.target-1`
- Check the disk is connected by `iscsicli reporttargetmappings`

9. initialize the connected disk with GPT data:
- Download (gdisk)[https://sourceforge.net/projects/gptfdisk/] 
- run gdisk\gdisk64, for disk number that is connected by iscsi
- type 'n' to create new partition
- specify '32M' for partition offset
- Divide "main dev size value" by 1024 to get value in KB, specify it for partition size like '+NNNK' , e.g. '+8351744K'
- for other partition parameters use defaults
- type 'w' to write partition table to disk
- After that a partition with data from original volume should appear
10. To end the session (and disconnect the disk):
- `iscsicli logouttarget <session-id>`, where <session-id>  is reported by `iscsicli reporttargetmappings`, e.g. `iscsicli logouttarget fffffa800626e018-4000013700000006`



document/ChangeLog

This file is part of pyTarget.
Copyright 2009-2010 Wu.Qing-xiu Wu.QX@hotmail.com

The pyTarget is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License (either version
2.1 of the License, or, at your option, any later version) and the GNU 
General Public License as published by the Free Software Foundation. See
the GNU Lesser General Public License for more details.

##############################################################################
 pyTarget-1.5.7.1:
------------------------------------------------------------------------------
    2011-4-1:
        Add Joshua's patch for isns:
        1) It adds a DevAttrQry() method to iSNSClient() to send the DevAttrQry request.
        2) Fix Task() method.

 pyTarget-1.5.7:
------------------------------------------------------------------------------
    2010-12-14:
        1. Fix text request negotiation.
        2. Fix some lock bug.
        3. Add Data-out data_sn check, etc

 pyTarget-1.5.6:
------------------------------------------------------------------------------
    2010-11-21:
        1. Modify file lock.
        2. Change some codes.

 pyTarget-1.5.4:
------------------------------------------------------------------------------
    2010-06-20:
        1. Add isns(scn, esi).
        2. Fix recovery r2t bug.

 pyTarget-1.5.3:
------------------------------------------------------------------------------
    2010-05-14:
        1. Fix iscsi scsi cmd_sn out of range.
        2. Add protect scsi device.

 pyTarget-1.5.2:
------------------------------------------------------------------------------
    2010-05-10:
        1. Fix some iscsi tape bugs.

 pyTarget-1.5.1:
------------------------------------------------------------------------------
    2010-04-12:
        1. add virtual tape library module.
        2. add VTL-driver for windows
        3. add iscsi initiator
        4. fix somes iscsi/scsi bugs

 pyTarget-1.0.6:
------------------------------------------------------------------------------
    2010-02-11:
        1. add snack data/r2t/data_ack/r_data.
        2. add all kinds of task management function.
        3. add iscsi error recover(ERL=2).
           support within command error recovery(e.g. digest error)
           support within connection error recovery(e.g. command retry)
           support connection recovery(e.g. task reassign)
           support session recovery

pyTarget-1.0.2:
------------------------------------------------------------------------------
    2009-11-27:
        1. add webui
        2. fix some iscsi compatible issues.
        3. support ses protocol.
        4. support scsi simulator.

pyTarget-0.16.0:
------------------------------------------------------------------------------
    2009-10-27:
        1. add cdb6, cdb10, cdb12, cdb16 SCSI command into scsi_lib
        2. add scsi error simulator for simulating scsi test
        3. add disk plist & glist into lun,
           support scsi reassign_blocks & read_defect_list scsi command.

pyTarget-0.15.0:
------------------------------------------------------------------------------
    2009-10-15:
        1. add initiator and target mutual chap
        2. fix InitiatorName and TargetName check while login phase
        3. main.py renamed pyTarget.py
        3. build all in one packed for windows
    2009-10-17:
        1. lun add lock for multi-connect
        2. modify cmd_cache for write cmd

pyTarget-0.13.1:
------------------------------------------------------------------------------
    2009-10-1:
        1. Add negotiation parameter into xml config file
        2. Fix Lun open flags

pyTarget-0.13:
------------------------------------------------------------------------------
    2009-9-28:
        1. Support scsi cdb 16 command
        2. Modify debug API
        3. Support multi-R2T solicited request.
        4. Simplified config.xml file
        5. Modify solicited data and unsolicited data.
       
    2009-9-29:
        1. Add command cache for multi-I/O
        2. Support mult-command
        3. Update Write/Read command

pyTarget-0.1:
------------------------------------------------------------------------------
    first version:
    1. Spuuort multi-target
    2. Support multi-session
    3. Support multi-connect/session (MC/s)
    4. Add CHAP (only support md5 chap)
    5. Support header & data digest
    6. Support ErrorRecoveryLevel = 0


src/tagt/main.py
#
#    iscsi target main function.
#    
#    Modify history:    
#    -----------------------------------------------------------------
#    Create by Wu.Qing-xiu (2009-09-27)
#

import os, comm.debug as debug
from iscsi.iscsi_sock import IscsiSock
from tagt.target import Target, TargetPool
from tagt.host import Host
from tagt.connect import Connect
from scsi.scsi_dev import Lun
from scsi.scsi_disk import Disk
from scsi.scsi_tape import Tape
from scsi import scsi_proto as sp
from scsi.scsi_enclosure import NEW_Enclosure
from config import read_config
from comm.debug import *
from comm.dev_file import DevFile
import platform

# for media type mask (changes the behaviour of shared media)
# TODO: move to media.factory file
SOURCE_MEDIA_FILE                        = 0x0 
SOURCE_MEDIA_WINDEV_MASK                 = 0xF000
SOURCE_MEDIA_WINDRIVE                    = 0x1000
SOURCE_MEDIA_WINVOL                      = 0x2000
SOURCE_MEDIA_WINVOL_EMULATED_DISK_LAYOUT = 0x4000

#
# target pool
#
TM = TargetPool()

#
# returns DevFile object based on target media
# and lun xml description
#
def _create_target_media(l):
    windev = None
    if (l.media & SOURCE_MEDIA_WINDEV_MASK) > 0:
        if platform.system() == "Windows":
            import media.win_dev
            windev = media.win_dev.WinDev(l.path , l.media == SOURCE_MEDIA_WINDRIVE, True)
            if l.media == SOURCE_MEDIA_WINVOL_EMULATED_DISK_LAYOUT:
                import media.emulated_layout_dev
                #TODO: check parms format, return error
                filenames = l.parms.split(';')
                #TODO: check files are initialized
                header = DevFile(filenames[0])
                footer = DevFile(filenames[1])
                emulated_gpt = media.emulated_layout_dev.EmulatedLayoutDev(windev, header, footer)
                windev = emulated_gpt
        else:
            DBG_WRN('Windows device type specified for non-Windows system')
    if windev:
        d = windev
    else:
        d = DevFile(l.path) # generic file media
    d.dev_lock()
    tp = l.type & (~sp.TYPE_PROTECT_MASK)
    if tp == sp.TYPE_DISK:  ln = Disk(l.id, l.cap, d, l.physical_sector, l.logical_sector)
    elif tp == sp.TYPE_TAPE:    ln = Tape(l.id, l.cap, d)
    elif tp == sp.TYPE_ENCLOSURE:   ln = NEW_Enclosure(l.id, d)
    else:ln = Lun(l.id, l.type, d)
    return ln

#
# config target
#
def read_target_config(config):
    '''
    Read configure and build target
    '''
    # update console debug level
    debug.CURRENT_DEBUG_LEVEL = config.debug_level

    # add target/host/lun
    for t in config.target_list:
        tg = Target(t.name, t.ip, t.port, t.portal)
        tg.config = t.para
        TM.add_target(tg)

        # scan target's host
        for h in t.host:
            ht = Host(h.name, h.pwd_t, h.pwd_i)
            tg.add_host(ht)
            
            # scan host's lun
            for l in h.lun:
                ln = _create_target_media(l)
                ln.initial()
                ht.add_lun(ln)
                # protect type
                if l.type & sp.TYPE_PROTECT_MASK:
                    ln.set_protect()
    DBG_PRN('configure target done')


def Run(ip, port):
    '''
    Start to run
    '''
    DBG_INF('initialize finish, start to run ...')

    sock = IscsiSock(ip, port)
    if sock.initial() == False:
        DBG_WRN('iscsi service start FAILED, check if port(%d) has been used' % port)
        return

    while True:
        if sock.accept() == False:
            DBG_WRN('accept a client FAILED')
            return

        #
        # Create a new thread to handle it.
        # 
        conn = Connect(sock)
        threading.Thread(None, conn.Start).start()

###################################################################
#                             Main
###################################################################

def iscsi_target_service(isStop = None):
    DBG_INF('initialize virtual device, please wait...')
    config = read_config('config.xml')
    if config:
        read_target_config(config)
        Run(config.ip, config.port)